org.eclipse.jdt.ui/ui/org/eclipse/refactoring/MakeStaticRefactoring.java

```diff
@@ -0,0 +1,885 @@
+
+		//Add static modifier to methodDeclaration
+		ModifierRewrite modRewrite= ModifierRewrite.create(rewrite, fTargetMethodDeclaration);
+		modRewrite.setModifiers(fTargetMethodDeclaration.getModifiers() | Modifier.STATIC, null);
+
+		//If fMethodDeclaraion hasInstanceUsages, adding a new parameter will be necessary
+		List<SingleVariableDeclaration> alreadyUsedParameters= fTargetMethodDeclaration.parameters();
+		String className= ((TypeDeclaration) fTargetMethodDeclaration.getParent()).getName().toString();
+		String paramName= generateUniqueParameterName(className, alreadyUsedParameters);
+
+		//Change instance Usages ("this" and "super") to paramName and set fHasInstanceUsage flag
+		fTargetMethodDeclaration.getBody().accept(new ChangeInstanceUsagesInMethodBody(paramName, rewrite, ast, status));
+
+		//Refactored method could unintentionally hide method of parent class
+		if (!fHasInstanceUsages && isOverriding(fTargetMethod.getDeclaringType())) {
+			status.merge(RefactoringStatus.createFatalErrorStatus(RefactoringCoreMessages.MakeStaticRefactoring_hiding_method_of_parent_type));
+			return;
+		}
+
+		IType parentType= fTargetMethod.getDeclaringType();
+		ITypeParameter[] classTypeParameters= parentType.getTypeParameters();
+
+		//Adding an instance parameter to the newly static method to ensure it can still access class-level state and behavior.
+		if (fHasInstanceUsages) {
+			addInstanceAsParamIfUsed(status, ast, rewrite, alreadyUsedParameters, className, paramName, classTypeParameters);
+		}
+
+		//Updates typeParamList of MethodDeclaration and inserts new typeParams to JavaDoc
+		updateMethodParameterTypes(status, ast, rewrite, classTypeParameters);
+
+		//A static method can't have override annotations
+		deleteOverrideAnnotation(rewrite);
+
+		//Changes can't be applied to directly to AST, edits are saved in fChangeManager
+		TextEdit methodDeclarationEdit= rewrite.rewriteAST();
+		addEditToChangeManager(methodDeclarationEdit, fTargetMethod.getCompilationUnit());
+	}
+
+	private String generateUniqueParameterName(String className, List<SingleVariableDeclaration> parameters) {
+		String classNameFirstLowerCase= Character.toLowerCase(className.charAt(0)) + className.substring(1); //makes first char lower to match name conventions
+		if (parameters == null || parameters.isEmpty())
+			return classNameFirstLowerCase;
+
+		boolean duplicateExists= false;
+		String combinedName= classNameFirstLowerCase;
+		int counter= 2;
+		while (true) {
+			for (SingleVariableDeclaration param : parameters) {
+				String paramString= param.getName().getIdentifier();
+				if (!(combinedName.equals(paramString))) {
+					duplicateExists= false;
+				} else {
+					duplicateExists= true;
+					break;
+				}
+			}
+			if (duplicateExists) {
+				combinedName= classNameFirstLowerCase + counter++;
+			} else {
+				return combinedName;
+			}
+		}
+	}
+
+	private final class ChangeInstanceUsagesInMethodBody extends ASTVisitor {
```

Comment:
Instead of having this class modify the outer class (in particular the `fHasInstanceUsages` field), better store the result in a field of this inner class and make the `modifyMethodDeclaration` write that field back to the outer class after applying this visitor. Then you can also move this inner class to a standalone class, which one can read and comprehend in isolation.

Reply:
Done.